### DISCUSSION TITLE
Using both SQL-Alchemy Base declarative and SQL Model

### DISCUSSION THREAD

**Discussion by jossefaz** [sqlalchemy/sqlalchemy#7374] [Usage Questions]
In one of our project, we used SQL Alchemy declarative Base for all of our models for years (and there is many). We want to try the new [SQLModel](https://sqlmodel.tiangolo.com/) library for our latest model declaration.

For that, we tried to declare it separately from the Base object, and call the `create_all` methods for both.

i.e : [`Base.metadata.create_all()`](https://docs.sqlalchemy.org/en/14/core/metadata.html#sqlalchemy.schema.MetaData.create_all) and [`SQLModel.metadata.create_all()`](https://sqlmodel.tiangolo.com/tutorial/create-db-and-table/?h=create_all#calling-create_all).

But the model declared with SQLModel does not recognize the table declared with the Base.

And at this moment, we cannot change all previous models declaration from Base to SQLModel.

Steps (see reproducible code sample above) :

- Create a Base declarative object from SQL Alchemy
- Import SQLModel
- Create a table that inherits from Base
- Create a table that inherits from SQLModel with table=True argument
- Define a foreign key from the last table to the first one
- call both create_all from Base.metadata and SQLModel.metadata objects

Here is a reproducible code :

```python
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy import Column
from sqlalchemy import Integer
from typing import Optional
from sqlmodel import Field, SQLModel
from sqlalchemy import String

# Declarative base object
Base = declarative_base()

class DummySATable(Base):
    __tablename__ = 'dummy_table'
    id = Column(Integer, primary_key=True, autoincrement=True)
    name = Column(String(32))

class DummyModelTable(SQLModel, table=True):
    id: Optional[int] = Field(default=None, primary_key=True)
    dummy_table_id : int = Field(default=None, foreign_key='dummy_table.id')
    name: str

Base.metadata.create_all(engine)
SQLModel.metadata.create_all(engine)
```

Here is the traceback (as we can see, the error is raised by SQLAlchemy and not by SQL Model, this is the raison why I open a discussion here) : 

```python
NoReferencedTableError                    Traceback (most recent call last)

/tmp/ipykernel_307893/3665898561.py in <module>
     24 
     25 Base.metadata.create_all(engine)
---> 26 SQLModel.metadata.create_all(engine)
     27 

~/project/venv/lib/python3.9/site-packages/sqlalchemy/sql/schema.py in create_all(self, bind, tables, checkfirst)
   4783         if bind is None:
   4784             bind = _bind_or_error(self)
-> 4785         bind._run_ddl_visitor(
   4786             ddl.SchemaGenerator, self, checkfirst=checkfirst, tables=tables
   4787         )

~/project/venv/lib/python3.9/site-packages/sqlalchemy/engine/base.py in _run_ddl_visitor(self, visitorcallable, element, **kwargs)
   3108     def _run_ddl_visitor(self, visitorcallable, element, **kwargs):
   3109         with self.begin() as conn:
-> 3110             conn._run_ddl_visitor(visitorcallable, element, **kwargs)
   3111 
   3112     @util.deprecated_20(

~/project/venv/lib/python3.9/site-packages/sqlalchemy/engine/base.py in _run_ddl_visitor(self, visitorcallable, element, **kwargs)
   2111 
   2112         """
-> 2113         visitorcallable(self.dialect, self, **kwargs).traverse_single(element)
   2114 
   2115     @util.deprecated(

~/project/venv/lib/python3.9/site-packages/sqlalchemy/sql/visitors.py in traverse_single(self, obj, **kw)
    522             meth = getattr(v, "visit_%s" % obj.__visit_name__, None)
    523             if meth:
--> 524                 return meth(obj, **kw)
    525 
    526     def iterate(self, obj):

~/project/venv/lib/python3.9/site-packages/sqlalchemy/sql/ddl.py in visit_metadata(self, metadata)
    820             tables = list(metadata.tables.values())
    821 
--> 822         collection = sort_tables_and_constraints(
    823             [t for t in tables if self._can_create_table(t)]
    824         )

~/project/venv/lib/python3.9/site-packages/sqlalchemy/sql/ddl.py in sort_tables_and_constraints(tables, filter_fn, extra_dependencies, _warn_for_cycles)
   1284                     continue
   1285 
-> 1286             dependent_on = fkc.referred_table
   1287             if dependent_on is not table:
   1288                 mutable_dependencies.add((dependent_on, table))

~/project/venv/lib/python3.9/site-packages/sqlalchemy/sql/schema.py in referred_table(self)
   3703 
   3704         """
-> 3705         return self.elements[0].column.table
   3706 
   3707     def _validate_dest_table(self, table):

~/project/venv/lib/python3.9/site-packages/sqlalchemy/util/langhelpers.py in __get__(self, obj, cls)
   1111         if obj is None:
   1112             return self
-> 1113         obj.__dict__[self.__name__] = result = self.fget(obj)
   1114         return result
   1115 

~/project/venv/lib/python3.9/site-packages/sqlalchemy/sql/schema.py in column(self)
   2408 
   2409             if tablekey not in parenttable.metadata:
-> 2410                 raise exc.NoReferencedTableError(
   2411                     "Foreign key associated with column '%s' could not find "
   2412                     "table '%s' with which to generate a "

NoReferencedTableError: Foreign key associated with column 'dummymodeltable.dummy_table_id' could not find table 'dummy_table' with which to generate a foreign key to target column 'id'
```

**Comment by CaselIT** [comment]
Hi,

You are trying to use two separate metadata, but you seem to want only one. You can pass a metadata object to declarative base. Don't know how you can specify a metadata in sqlmodel.

**Reply by jossefaz** [reply]
From the docs I did not found that an engine can have only one metadata object...that's why I tried this way, I thought that it would just update the current metadata. 

Can I merge two metadata object into one somehow, like update the current metadata with another ?

**Reply by CaselIT** [reply]
An engine does not have any notion of a metadata, so that's not the issue.
Here you are trying to reference two tables that are in two different metadata and I don't think that's supported.

You may try passing `checkfirst=false` to create, but it will then error if the table exists

**Reply by jossefaz** [reply]
So what would be the way of "combining" those two metadata declarations?

**Reply by CaselIT** [reply]
I don't think it's possible to combine two metadata.

As suggested above you should share them from the beginning

**Reply by jossefaz** [reply]
I think I've found :
Doing this seems to do the trick
```python
Base = declarative_base()
SQLModel.metadata = Base.metadata
```

**Reply by jossefaz** [reply]
@CaselIT What do you think about this possible solution? Is there any points that I need to be aware of?

**Reply by CaselIT** [reply]
Seems to be ok.

You can also pass metadata as an argument to declarative_base if that does not work.

Just note that SQLModel will create an index for every column by default

**Reply by zzzeek** [reply]
> Just note that SQLModel will create an index for every column by default

what, why?  that would be crazy?

**Reply by CaselIT** [reply]
No clue, it seems a bug.
I tried opening a pr to fix https://github.com/tiangolo/sqlmodel/pull/11 but sqlmodel does not seem to be particularly maintained

**Comment by zzzeek** [comment]
Hi -

why is this not being reported *first* to SQLModel?   I am sure @tiangolo  has ideas on this which are specific to the implementation of SQLModel, he knows it best and he also has a deep understanding of SQLAlchemy's declarative base.  On this end, we only know the latter.   Why report it here ?
